Updated deployment functions

ccaum opened the pull request:
This PR introduces two new functions for use in custom deployment policies. It also modified an existing function to make use of a new one under the hood. 

-  **cd4pe_deployments::get_impact_analysis_reports()**  — NEW — This function takes a commit and a repository name as parameters will return every Impact Analysis report run against the commit previously in the pipeline. I THINK this will not return past Impact Analysis reports from previous pipeline runs on the same commit (i.e. if the same pipeline is manually triggered multiple times with the same commit), but I cannot test this in my environment

- **cd4pe_deployments::get_pipeline_trigger_events** — NEW —This function is similar to pre-existing get_pipeline_trigger_event function, however this one will return all trigger events for a given pipeline. The get_pipeline_trigger_event function will only return the last trigger event on a pipeline.

- **cd4pe_deployments::get_pipeline_trigger_event** — MODIFIED — This function has been updated to call the cd4pe_deployments::get_pipeline_trigger_events() function and return the first result found. This function should be deprecated in the future, in favor of just providing the get_pipeline_trigger_events() function, however it needs to stay around for now for backwards compatibility.

ccaum committed:
Prevent repeat deployment approval pending states

Previous to this commit, the cd4pe_deployments::wait_for_approval() function
would immediately attend to put the deployment approval into a pending
state. However, it's possible for the approval to already be approved
earlier in the deployment plan, before the wait_for_approval() function
is called.

This commit first checks the status of the deployment approval and
returns with the status if it's already approved or denied

ccaum committed:
Add new funcitons to simplify geting impact analysis reports

ccaum committed:
Remove erroneuous script

ccaum committed:
Fix how the get_pipeline_trigger_event funciton calls its dependency

ccaum committed:
Make rubocop less angry

ccaum committed:
Make rubocup less angry

ccaum committed:
Robocop is really pedantic

VitaC123 commented on lib/puppet/functions/cd4pe_deployments/get_pipeline_trigger_event.rb:
```diff
@@ -24,18 +24,16 @@
 
   def get_pipeline_trigger_event(repo_name, pipeline_id, commit_sha)
-    client = PuppetX::Puppetlabs::CD4PEClient.new
+    results = call_function('cd4pe_deployments::get_pipeline_trigger_events', repo_name, pipeline_id, commit_sha)
```
I see the method already exists, but do we want to expose this information as "trigger events" to our customers? I wonder if we might want to label this method as something more self-descriptive, like `getPipelineEvent` or `getTimelineEvent`. Just throwing the idea out there, not sure it's worth the breaking change.

ccaum replied:
I'd prefer to deprecate this function, but we can't introduce a backwards breaking change until 4.0. Not sure how our deprecation process works. And maybe we don't need to worry about it since it's still in beta?

Ziaunys replied:
I think "pipeline run events" makes more sense, but we might want to just stick with "trigger events" for the sake of consistency.

VitaC123 commented on lib/puppet/functions/cd4pe_deployments/get_pipeline_trigger_events.rb:
```diff
@@ -0,0 +1,41 @@
+require 'puppet_x/puppetlabs/cd4pe_function_result'
+
+# @summary Search for all the pipeline triggers that have occurred with a given SHA
+Puppet::Functions.create_function(:'cd4pe_deployments::get_pipeline_trigger_events') do
```
Admittedly pedantic, (but hopefully not as bad as Robocop), could this be a `list` method? Otherwise, I have to parse for that "s" just to know it's doing something different.

VitaC123 commented on lib/puppet/functions/cd4pe_deployments/get_impact_analysis_reports.rb:
```diff
@@ -0,0 +1,55 @@
+    required_param 'String', :repo_name
+  end
+
+  def get_impact_analysis_reports(commit_sha, repo_name)
```
I love how you've leveraged existing codez to get the reports. 🎉 

Not that we necessarily want to tackle this now, but I'm curious if you or others have thoughts about recreating this as a new endpoint in the Java codebase to allow this client to make one single network call, and to reduce what this client needs to know about CD4PE to get the information it needs. If we did want to do that, we could replace the guts of this ruby code with a single network call, later.

ccaum replied:
That would be nice.

Ziaunys replied:
I agree. This is a really good way to quickly get new functionality in to custom deployment policies, but in the long term we shouldn't be adding too much logic in the client.

Ziaunys commented on lib/puppet/functions/cd4pe_deployments/get_pipeline_trigger_event.rb:
```diff
@@ -24,18 +24,16 @@
   end
 
   def get_pipeline_trigger_event(repo_name, pipeline_id, commit_sha)
```
You mention deprecating this method, but @kreeuwijk added this in https://github.com/puppetlabs/puppetlabs-cd4pe_deployments/pull/49 for a specific reason so you might want to check in with him first. I see some value in having a method to get the latest one vs. all of them. We could rename the function though.

kreeuwijk replied:
I'm fine with this, I looked at the code changes and this doesn't break anything. I'm happy to switch to `get_pipeline_trigger_events()` and do the `.first` myself in the plan code .

Ziaunys commented on lib/puppet/functions/cd4pe_deployments/get_impact_analysis_reports.rb:
```diff
@@ -0,0 +1,55 @@
+require 'puppet_x/puppetlabs/cd4pe_function_result'
+
+# @summary Get all the Impact Analysis reports for a commit in the current pipeline
+Puppet::Functions.create_function(:'cd4pe_deployments::get_impact_analysis_reports') do
```
Perhaps we should consider adding unit tests for new functions.

ccaum committed:
Ensure approval state before entering wait loop

This commit ensures the deployment is not already in a pending,
approved, or declined state before it sets the deployment to pending.
The tests have been updated to stub out the new extra call to get the
deployment state

ccaum committed:
Appeasing the robucop god

ccaum committed:
Add comma because rubocop

ccaum committed:
Merge branch 'fix_pending_if_already_approved' into updated_deployment_functions

* fix_pending_if_already_approved:
  Add comma because rubocop
  Appeasing the robucop god
  Ensure approval state before entering wait loop
  Make rubocup less angry
  Prevent repeat deployment approval pending states

ccaum committed:
The search_pipeline function doesn't return a result

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/puppetlabs/puppetlabs-cd4pe_deployments?pullRequest=52) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/puppetlabs/puppetlabs-cd4pe_deployments?pullRequest=52) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/puppetlabs/puppetlabs-cd4pe_deployments?pullRequest=52) it.</sub>

eputnam commented:
closing this because it's old
